File: lib/plugins/rdbms_corparch/registry/parser.py
# Copyright (c) 2018 Charles University, Faculty of Arts,
#                    Institute of the Czech National Corpus
# Copyright (c) 2018 Tomas Machalek <tomas.machalek@gmail.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; version 2
# dated June, 1991.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

import re
from functools import wraps
from plugins.rdbms_corparch.registry import Attribute, PosAttribute, Struct, SimpleAttr, RegistryConf


DEBUG = 0


class RegistrySyntaxError(Exception):
    pass


def infer_encoding(file_path):
    with open(file_path) as fr:
        for line in fr:
            if 'ENCODING' in line:
                if 'utf8' in line.lower() or 'utf-8' in line.lower():
                    return 'utf-8'
                elif 'iso' in line.lower() and '8859-2' in line:
                    return 'iso-8859-2'
                break
    return 'utf-8'


def watchable(f):

    @wraps(f)
    def fn(slf, token, obj):
        if DEBUG:
            print(('fn: {0}, tok: {1}, curr: {2}'.format(f.__name__, token if len(token) > 0 else '',
                                                         obj.__class__.__name__ if obj else None)))
        return f(slf, token, obj)
    return fn


class Token(object):

    def __init__(self, type, value=None):
        self._type = type
        self._value = value


class Tokenizer(object):

    def __init__(self, infile, encoding):
        self._fr = infile
        self._encoding = encoding

    def __call__(self):
        ans = []
        for line in self._fr:
            line = line.decode(self._encoding)
            items = re.split(r'\s+', line)
            line_ans = []
            is_q = False
            for item in items:
                if item == '':
                    continue
                if item[0] == '"' and item[-1] == '"' and len(item) > 1:
                    line_ans.append(item[1:-1])
                elif item[0] == '"' and not is_q:
                    line_ans.append([])
                    is_q = True
                    v = item[1:]
                    line_ans[-1].append(v)
                elif item[-1] == '"':
                    is_q = False
                    v = item[:-1]
                    if type(line_ans[-1]) is list and line_ans[-1][0] == '"' and v == '"':
                        line_ans[-1].append(' ')
                    else:
                        line_ans[-1].append(v)
                else:
                    if is_q:
                        line_ans[-1].append(item)
                    else:
                        line_ans.append(item)
            tmp = [' '.join(v) if type(v) is list else v for v in line_ans] + ['$']
            if len(tmp) > 0:
                ans.append(tmp)
        return [v for subl in ans for v in subl]


class Parser(object):

    def __init__(self, corpus_id, variant, tokens, backend):
        self._tokens = tokens
        self._items = RegistryConf(corpus_id, variant, backend)
        self._posattr_idx = 0

    @staticmethod
    def is_key(s):
        return re.match(r'[A-Z]+', s)

    @staticmethod
    def is_value(s):
        return s != '{' and s != '}' and s != '#'

    @watchable
    def state_0(self, token, obj):
        if token == '$':
            return self.state_0, obj
        elif token.startswith('#'):
            return self.state_3, obj
        elif self.is_key(token):
            if obj:
                self._items.add_item(obj)
            if token == 'ATTRIBUTE':
                attr = PosAttribute(position=self._posattr_idx, name=token)
                self._posattr_idx += 1
                return self.state_1, attr
            elif token == 'STRUCTURE':
                return self.state_1, Struct(token)
            else:
                return self.state_1, SimpleAttr(token)
        else:
            raise RegistrySyntaxError('in state 0 cannot process: {0}'.format(token))

    @watchable
    def state_1(self, token, obj):
        if self.is_value(token):
            if isinstance(obj, SimpleAttr):
                obj.value = token
            else:
                obj.name = token
            return self.state_2, obj
        else:
            raise RegistrySyntaxError('in state 1 cannot process: {0}'.format(token))

    @watchable
    def state_2(self, token, obj):
        if token == '$':
            return self.state_0, obj
        elif token == '{':
            return self.state_4, obj

    @watchable
    def state_3(self, token, obj):
        if token == '$':
            return self.state_0, obj
        return self.state_3, obj

    def state_3b(self, token, obj):
        if token == '$':
            return self.state_4, obj
        return self.state_3b, obj

    def state_3c(self, token, obj):
        if token == '$':
            return self.state_7, obj
        return self.state_3c, obj

    @watchable
    def state_4(self, token, obj):
        if token == '}':
            return self.state_0, obj
        elif self.is_key(token):
            if token == 'ATTRIBUTE':
                obj.new_item(Attribute())
            else:
                obj.new_item(SimpleAttr(token))
            return self.state_5, obj
        elif token == '$':
            return self.state_4, obj
        elif token.startswith('#'):
            return self.state_3b, obj

    @watchable
    def state_5(self, token, obj):
        if self.is_value(token):
            if isinstance(obj.last_item, Attribute):
                obj.last_item.name = token
            else:
                obj.last_item.value = token
            return self.state_6, obj

    @watchable
    def state_6(self, token, obj):
        if token == '$':
            return self.state_4, obj
        elif token == '{':
            return self.state_7, obj

    @watchable
    def state_7(self, token, obj):
        if self.is_key(token):
            obj.last_item.new_item(SimpleAttr(token))
            return self.state_8, obj
        elif token == '$':
            return self.state_7, obj
        elif token == '}':
            return self.state_4, obj
        elif token.startswith('#'):
            return self.state_3c, obj

    @watchable
    def state_8(self, token, obj):
        if self.is_value(token):
            obj.last_item.last_item.value = token
            return self.state_9, obj

    @watchable
    def state_9(self, token, obj):
        if token == '$':
            return self.state_7, obj

    def __call__(self):
        i = 0
        fn = self.state_0
        obj = None
        while fn is not None and i < len(self._tokens):
            fn, obj = fn(self._tokens[i], obj)
            i += 1
        if obj:
            self._items.add_item(obj)
        return self._items

File: public/files/js/models/cqleditor/parser.ts
/*
 * Copyright (c) 2018 Charles University in Prague, Faculty of Arts,
 *                    Institute of the Czech National Corpus
 * Copyright (c) 2018 Tomas Machalek <tomas.machalek@gmail.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 2
 * dated June, 1991.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import * as Kontext from '../../types/kontext';
import { parse as parseQuery, SyntaxError } from 'cqlParser/parser';
import { IAttrHelper, NullAttrHelper } from './attrs';
import { tuple } from 'cnc-tskit';
import { ParsedAttr, ParsedPQItem, RuleCharMap } from './rules';

export type { ParsedAttr, ParsedPQItem } from './rules';


/**
 * ParserStack is used along with PEG parser to walk
 * through rules (enter, exit, fail) and insert rule
 * marks to the query via RuleCharMap instance.
 */
class ParserStack {

    private stack:Array<{rule:string; lastPos:number}>;

    private rcMap:RuleCharMap;

    private lastPos:number;


    constructor(rcMap:RuleCharMap) {
        this.stack = [];
        this.rcMap = rcMap;
        this.lastPos = 0;
    }

    push(rule:string, lastPos:number):void {
        this.stack.push({
            rule: rule,
            lastPos: lastPos
        });
    }

    pop(lastPos:number):void {
        const v = this.stack.pop();
        if (this.isTerminal(v.rule)) {
            if (v.lastPos < lastPos) {
                this.rcMap.set(v.lastPos, lastPos, v.rule);
                this.lastPos = Math.max(this.lastPos, lastPos);
            }

        } else {
            this.rcMap.addNonTerminal(v.lastPos, lastPos, v.rule);
        }
    }

    popFail(rule:string, lastPos:number):void {
        this.stack.pop();
    }

    private isTerminal(rule:string):boolean {
        return !!/^[A-Z_]+$/.exec(rule);
    }

    getLastPos():number {
        return this.lastPos;
    }
}


const escapeQuery = (v:string):string => {
    return v.replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;');
};


interface HSArgs {
    query:string;
    applyRules:Array<string>;
    he:Kontext.Translator;
    ignoreErrors:boolean;
    attrHelper:IAttrHelper;
    parserRecoverIdx:number;
    wrapLongQuery:boolean;
    wrapRange:(startIdx:number, endIdx:number)=>[string, string];
}

function _highlightSyntax({
    query,
    applyRules,
    he,
    ignoreErrors,
    attrHelper,
    parserRecoverIdx,
    wrapLongQuery,
    wrapRange
}:HSArgs):[string, Array<ParsedAttr>, Array<ParsedPQItem>, string] {

    const rcMap = new RuleCharMap(query, he, attrHelper, wrapLongQuery, wrapRange);
    const stack = new ParserStack(rcMap);

    const wrapUnrecognizedPart = (v:string, numParserRecover:number, error:SyntaxError):[string, string|undefined] => {
        if (numParserRecover === 0 && error) {
            const title = he.translate(
                'query__unrecognized_input_{wrongChar}{position}',
                {
                    wrongChar: error.found,
                    position: error.location.start.column
                }
            );
            const style = 'text-decoration: underline dotted red';
            return tuple(
                `<span title="${escapeQuery(title)}" style="${style}">` + escapeQuery(v) + '</span>',
                title
            );

        }
        return tuple(escapeQuery(v), undefined);
    }

    let parseError:SyntaxError = null;
    try {
        parseQuery(query, {
            startRule: applyRules[0],
            tracer: {
                trace: (v) => {
                    switch (v.type) {
                        case 'rule.enter':
                            stack.push(v.rule, v.location.start.offset);
                        break;
                        case 'rule.fail':
                            stack.popFail(v.rule, v.location.end.offset);
                        break;
                        case 'rule.match':
                            stack.pop(v.location.end.offset);
                        break;
                    }
                }
            }
        });

    } catch (e) {
        parseError = e;
        if (!ignoreErrors) {
            throw e;
        }
    }

    const lastPos = stack.getLastPos();
    const [ans, parsedAttrs, pqItems] = rcMap.generate();

    if (query.length === 0) {
        return tuple('', [], [], undefined);

    } else if (lastPos < query.length && applyRules.length > 1) {
        // try to apply a partial rule to the rest of the query
        const srch = /^([^\s]+|)(\s+)(.+)$/.exec(query.substr(lastPos));
        if (srch !== null) {
            const [partial, parsedAttrs] = _highlightSyntax({
                query: srch[3],
                applyRules: srch[1].trim() !== '' ? applyRules.slice(1) : applyRules,
                he: he,
                ignoreErrors: true,
                attrHelper,
                wrapLongQuery: false,
                wrapRange,
                parserRecoverIdx: parserRecoverIdx + 1
            });

            const [subQueryHighlight, err] = wrapUnrecognizedPart(srch[1] + srch[2], parserRecoverIdx, parseError);
            return tuple(
                ans + subQueryHighlight + partial,
                parsedAttrs,
                pqItems,
                err
            );
        }
    }
    const [subQueryHighlight, err] = wrapUnrecognizedPart(query.substr(lastPos), parserRecoverIdx, parseError);
    return tuple(
        ans + (query.substr(lastPos).length > 0 ? subQueryHighlight : ''),
        parsedAttrs,
        pqItems,
        query.length > lastPos ? err : undefined
    );
}

export function getApplyRules(querySuperType:Kontext.QuerySupertype):Array<string> {
    switch (querySuperType) {
        case 'pquery':
            return ['PQuery', 'Query', 'WithinContainingPart', 'Sequence', 'RegExpRaw'];
        case 'conc':
            return ['Query', 'WithinContainingPart', 'Sequence', 'RegExpRaw'];
        case 'wlist':
            return ['RegExpRaw'];
    }
}

/**
 * highlightSyntax generates a syntax highlighting
 * for a query by embedding an HTML markup into
 * the resulting string. It also provides a list
 * of detected attributes and structures and paradigmatic
 * query sub-queries.
 *
 * @return a 4-tuple (
 *   highlighted query,
 *   list of detected attributes,
 *   list of detected paradigm. query sub-queries
 *   syntax error
 * )
 */
export function highlightSyntax(
    {
        query, querySuperType, he, attrHelper, wrapRange
    }:{
        query:string,
        querySuperType:Kontext.QuerySupertype,
        he:Kontext.ComponentHelpers,
        attrHelper:IAttrHelper,
        wrapRange:((startIdx:number, endIdx:number)=>[string, string])|undefined,
    }):[string, Array<ParsedAttr>, Array<ParsedPQItem>, string] {

    return _highlightSyntax({
        query: query,
        applyRules: getApplyRules(querySuperType),
        he: he,
        ignoreErrors: true,
        attrHelper: attrHelper ? attrHelper : new NullAttrHelper(),
        wrapLongQuery: false,
        wrapRange,
        parserRecoverIdx: 0
    });
}

/**
 *
 * @return a 4-tuple (highlighted query, list of detected attributes, list of detected Paradigm. query blocks, syntax error)
 */
export function highlightSyntaxStatic(
    {query, querySuperType, he}:{
        query:string,
        querySuperType:Kontext.QuerySupertype,
        he:Kontext.Translator
    }):[string, Array<ParsedAttr>, Array<ParsedPQItem>, string] {

    return _highlightSyntax({
        query: query,
        applyRules: getApplyRules(querySuperType),
        he: he,
        ignoreErrors: true,
        attrHelper: new NullAttrHelper(),
        wrapLongQuery: true,
        wrapRange: undefined,
        parserRecoverIdx: 0
    });
}